### PR TITLE
Add tunnel_key action

### PR DESCRIPTION
**Pull request by dennisafa**
* Adds tunnel_key action functionality 
Apologies for the git noise, did a bad merge.

### CODE REVIEW

File: m_tunnel_key_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+				t.Fatalf("Unexpected error: %v", err2)
+			}
+			testcase.val.Tm = tm
+			if diff := cmp.Diff(val, testcase.val); diff != "" {
+				t.Fatalf("VLan missmatch (want +got):\n%s", diff)
+			}
+		})
+	}
+	t.Run("nil", func(t *testing.T) {
+		_, err := marshalVlan(nil)
```
Comment (florianl) [inline m_tunnel_key_test.go:55]: I guess this is a typo:
```suggestion
		_, err := marshalTunnelKey(nil)
```

File: m_tunnel_key_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+			err2 := unmarshalTunnelKey(newData, &val)
+			if err2 != nil {
+				if testcase.err2 != nil && testcase.err2.Error() == err2.Error() {
+					return
+				}
+				t.Fatalf("Unexpected error: %v", err2)
+			}
+			testcase.val.Tm = tm
+			if diff := cmp.Diff(val, testcase.val); diff != "" {
+				t.Fatalf("VLan missmatch (want +got):\n%s", diff)
```
Comment (florianl) [inline m_tunnel_key_test.go:50]:
```suggestion
				t.Fatalf("TunnelKey missmatch (want +got):\n%s", diff)
```

File: m_tunnel_key.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	tcaTunnelKeyEncTOS
+	tcaTunnelKeyEncTTL
+)
+
+// TunnelKey contains attribute of the TunnelKey discipline
+type TunnelKey struct {
+	Parms           *TunnelParms
+	Tm              *Tcft
+	KeyEncIPv4Src   *net.IP
+	KeyEncIPv4Dst   *net.IP
```
Comment (florianl) [inline m_tunnel_key.go:32]: Can you change this to `KeyEncSrc` and `KeyEncDst` and check in `marshalTunnelKey` if it is IPv4 or IPv6 and set the attribute accordingly?

File: m_tunnel_key_test.go (hunk 1 above)
Comment (dennisafa) [reply m_tunnel_key_test.go]: Yep, copied over the marshalVLan tests and forgot to change a few things...thanks!

File: m_tunnel_key.go (hunk 3 above)
Comment (dennisafa) [reply m_tunnel_key.go]: Done. I just converted the net.IP object to bytes - works for both IPv6 and IPv4.

File: m_tunnel_key.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+	"github.com/mdlayher/netlink"
+)
+
+const (
+	tcaTunnelUnspec = iota
+	tcaTunnelKeyTm
+	tcaTunnelKeyParms
+	tcaTunnelKeyEncSrc
+	tcaTunnelKeyEncDst
```
Comment (florianl) [inline m_tunnel_key.go:15]: These consts have to be in sync with https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/tc_act/tc_tunnel_key.h#L25
So to merge `KeyEncSrc` and `KeyEncDst` for the IPv6 and IPv4 case only works for the Go struct `TunnelKey` but you have to differentiate when communicating with the kernel.

File: m_tunnel_key.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	tcaTunnelKeyEncTTL
+)
+
+// TunnelKey contains attribute of the TunnelKey discipline
+type TunnelKey struct {
+	Parms           *TunnelParms
+	Tm              *Tcft
+	KeyEncSrc       *net.IP
+	KeyEncDst       *net.IP
+	KeyEncKeyID     *uint64
```
Comment (florianl) [inline m_tunnel_key.go:31]: `TCA_TUNNEL_KEY_ENC_KEY_ID` seems to be `u32`
https://elixir.bootlin.com/linux/latest/source/net/sched/act_tunnel_key.c#L338

**Review by florianl — CHANGES_REQUESTED** [review]

File: m_tunnel_key.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	if info.Parms != nil {
+		data, err := marshalStruct(info.Parms)
+		if err != nil {
+			return []byte{}, err
+		}
+		options = append(options, tcOption{Interpretation: vtBytes, Type: tcaTunnelKeyParms, Data: data})
+	}
+	if info.KeyEncSrc != nil {
```
Comment (florianl) [inline m_tunnel_key.go:64]: `info.KeyEncSrc` can either be IPv6 or IPv4. Using always `tcaTunnelKeyEncIPv4Src` might lead to interesting side effects.
And same for `...Dst`.

File: m_tunnel_key.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+// unmarshalTunnelKey parses the TunnelKey-encoded data and stores the result in the value pointed to by info.
+func unmarshalTunnelKey(data []byte, info *TunnelKey) error {
+	ad, err := netlink.NewAttributeDecoder(data)
+	if err != nil {
+		return err
+	}
+	ad.ByteOrder = nativeEndian
+	for ad.Next() {
+		switch ad.Type() {
```
Comment (florianl) [inline m_tunnel_key.go:99]: cases `tcaTunnelKeyEncIPv6Src` and `tcaTunnelKeyEncIPv6Dst` are missing.

File: nest.go [hunk 8]
```diff
@@ -87,6 +88,12 @@ func marshalAttributes(options []tcOption) ([]byte, error) {
 				return []byte{}, err
 			}
 			ad.Bytes(option.Type, data.Bytes())
+		case vtUint64Be:
```
Comment (florianl) [inline nest.go:91]: Is this still needed?

File: m_tunnel_key.go (hunk 6 above)
Comment (dennisafa) [reply m_tunnel_key.go:64]: I'll fix these issues - was doing some testing.

File: nest.go (hunk 8 above)
Comment (dennisafa) [reply nest.go]: Not anymore - removed

File: m_tunnel_key.go (hunk 7 above)
Comment (dennisafa) [reply m_tunnel_key.go:115]: Added the two cases

File: m_tunnel_key.go (hunk 6 above)
Comment (dennisafa) [reply m_tunnel_key.go:64]: Should be fixed now.

**Review by florianl — APPROVED** [review]
Thanks for addressing the comments and contributing!

**Comment by dennisafa** [comment]
> Thanks for addressing the comments and contributing!

no problem!

**Pull request merged**